azurelinuxagent/common/persist_firewall_rules.py

```diff
@@ -0,0 +1,219 @@
+#
+# Copyright 2018 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import os
+import sys
+
+import shutil
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.cgroupapi import CGroupsApi
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.utils import shellutil, fileutil
+from azurelinuxagent.common.utils.networkutil import AddFirewallRules
+from azurelinuxagent.common.utils.shellutil import CommandError
+
+
+class PersistFirewallRulesHandler(object):
+
+    __SERVICE_FILE_CONTENT = """
+# This drop-in unit file was created by the Azure VM Agent.
+# Do not edit.
+[Unit]
+Description=Setup network rules for WALinuxAgent 
+Before=network-pre.target
+Wants=network-pre.target
+DefaultDependencies=no
+ConditionPathExists={0}
+
+[Service]
+Type=oneshot
+Environment="DST_IP={1}" "UID={2}" "WAIT={3}"
+ExecStart={4} {0} --dst_ip=${{DST_IP}} --uid=${{UID}} ${{WAIT}}
+RemainAfterExit=false
+
+[Install]
+WantedBy=network.target
+"""
+
+    __OVERRIDE_CONTENT = """
+# This drop-in unit file was created by the Azure VM Agent.
+# Do not edit.
+[Service]
+Environment=
+Environment="DST_IP={0}" "UID={1}" "WAIT={2}"
+"""
+
+    _AGENT_NETWORK_SETUP_BIN_FILE = "waagent_network_setup.py"
+    _AGENT_NETWORK_SETUP_NAME_FORMAT = "{0}-network-setup.service"
+    _DROP_IN_ENV_FILE_NAME = "10-environment.conf"
+
+    _FIREWALLD_RUNNING_CMD = ["firewall-cmd", "--state"]
+
+    def __init__(self, dst_ip, uid):
+        """
+        If not, then create the new service file and copy it over to the /lib/systemd/system/ if doesnt exist
+          - Check if unit file already exists, if it does, leave it
+          - Else
+              - Dynamically create the unit file and setup the IPs correctly
+              - Check if /sbin/walinuxagent_setup_service.py exists
+              - If not, then copy it over
+          - Create a drop in conf file in /etc/systemd/system/waagent-network-setup.d/override.conf and override the ExecStart with the new values
+            - Do this everytime on service restart
+        """
+        osutil = get_osutil()
+        self._network_setup_service_name = self._AGENT_NETWORK_SETUP_NAME_FORMAT.format(osutil.get_service_name())
+        self._is_systemd = CGroupsApi.is_systemd()
```

Comment:
is_systemd probably fits better in osutil (it's weird that the firewall rules have a dependency on cgroups, for example). I'll move it there in a future PR.

Reply:
Yup agreed. I can do it but I dont have enough context on how to determine if systemd is present or not in multiple distros (or is the path /run/systemd/system standard for all?)

Reply:
That is the path that systemd itself checks. Didn't check whether it changes across systemd versions.

But leaving aside the implemtation, osutil would be a better place for it. I'll make the change as part of the work I am doing in cgroups.